src/adt/edgeset.c

```diff
@@ -871,3 +887,997 @@ edge_set_ordered_iter_next(struct edge_ordered_iter *eoi, struct fsm_edge *e)
+			} else {
+				if (SYMBOLS_GET(eg->symbols, it->j)) {
+					e->symbol = it->j;
+					e->state = eg->to;
+					it->j++;
+					return 1;
+				}
+				it->j++;
+			}
+		}
+
+		it->i++;
+		it->j = 0;
+	}
+
+	return 0;
+}
+
+void
+edge_set_rebase(struct edge_set **pset, fsm_state_t base)
+{
+	size_t i;
+	struct edge_set *set;
+
+	assert(pset != NULL);
+	set = *pset;
+
+#if LOG_BITSET
+	fprintf(stderr, " -- edge_set_rebase: set %p, base %d\n",
+	    (void *)set, base);
+#endif
+
+	if (edge_set_empty(set)) {
+		return;
+	}
+
+	for (i = 0; i < set->count; i++) {
+		struct edge_group *eg = &set->groups[i];
+		eg->to += base;
+	}
+}
+
+int
+edge_set_replace_state(struct edge_set **pset, const struct fsm_alloc *alloc,
+    fsm_state_t old, fsm_state_t new)
+{
+	size_t i;
+	struct edge_set *set;
+	struct edge_group cp;
+
+	assert(pset != NULL);
+	set = *pset;
+
+#if LOG_BITSET
+	fprintf(stderr, " -- edge_set_replace_state: set %p, state %d -> %d\n",
+	    (void *)set, old, new);
+#endif
+
+	if (edge_set_empty(set)) {
+		return 1;
+	}
+
+	/* Invariants: if a group with .to == old appears in the group,
+	 * it should only appear once. Replacing .to may lead to
+	 * duplicates, so duplicates may need to be merged after. */
+
+	for (i = 0; i < set->count; i++) {
+		struct edge_group *eg = &set->groups[i];
+		if (eg->to == old) {
+			const size_t to_mv = set->count - i;
+			unsigned char label;
+			if (!find_first_group_label(eg, &label)) {
+				return 1; /* ignore empty group */
+			}
+
+#if LOG_BITSET
+			fprintf(stderr, " -- edge_set_replace_state: removing group at %ld with .to=%d\n", i, old);
+#endif
+			/* Remove group */
+			memcpy(&cp, eg, sizeof(cp));
+			memmove(&set->groups[i],
+			    &set->groups[i + 1],
+			    to_mv * sizeof(set->groups[i]));
+			set->count--;
+
+#if LOG_BITSET
+			dump_edge_set(set);
+			fprintf(stderr, " -- edge_set_replace_state: reinserting group with .to=%d and label 0x%x\n", new, (unsigned)label);
+#endif
+
+			/* Realistically, this shouldn't fail, because
+			 * edge_set_add only fails on allocation failure
+			 * when it needs to grow the backing array, but
+			 * we're removing a group and then adding the
+			 * group again so add's bookkeeping puts the
+			 * group in the appropriate place. */
+			if (!edge_set_add(&set, alloc,
+				label, new)) {
+				return 0;
+			}
+			dump_edge_set(set);
+
+			for (i = 0; i < set->count; i++) {
+				eg = &set->groups[i];
+				if (eg->to == new) {
+					size_t w_i;
+#if LOG_BITSET
+					fprintf(stderr, " -- edge_set_replace_state: found new group at %ld, setting other labels from copy\n", i);
+#endif
+					for (w_i = 0; w_i < 4; w_i++) {
+						eg->symbols[w_i] |= cp.symbols[w_i];
+					}
+					dump_edge_set(set);
+					return 1;
+				}
+			}
+			assert(!"internal error: just added, but not found");
+		}
+	}
+	return 1;
+}
+
+int
+edge_set_empty(const struct edge_set *s)
+{
+	size_t i;
+	if (s == NULL || s->count == 0) {
+		return 1;
+	}
+
+	for (i = 0; i < s->count; i++) {
```

Comment:
This needs to account for groups whose labels have all been removed, since currently we do not prune empty groups. It's not likely to be a common situation in practice, so the performance hit is probably okay. Property testing found a couple cases where that decision led to internal invariants breaking, all have been fixed.

Reply:
Do you know when that can happen?

Reply:
You could ask theft by commenting out that part, adding `return 0;`, and running `build/theft/theft -n adt_edge_set`, it should produce at least two different kinds of failures in well under a second. :)